### PR TITLE
Add Ord type class

### PR DESCRIPTION
**Pull request by esbenbjerre**
This PR provides
- [x] intrinsics for `<`, `<=`, `>` and `>=`
- [x] Ord type class

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Does this mean that we can remove BinaryDeprecated? :)

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 1]
```diff
@@ -428,6 +428,22 @@ object Simplifier extends Phase[TypedAst.Root, SimplifiedAst.Root] {
...
+
+          case SemanticOperator.CharOp.Gt | SemanticOperator.Float32Op.Gt | SemanticOperator.Float64Op.Gt
+               | SemanticOperator.Int8Op.Gt | SemanticOperator.Int16Op.Gt | SemanticOperator.Int32Op.Gt
+               | SemanticOperator.Int64Op.Gt | SemanticOperator.BigIntOp.Gt => BinaryOperator.Greater
+
+          case SemanticOperator.CharOp.Ge | SemanticOperator.Float32Op.Ge | SemanticOperator.Float64Op.Ge
+               | SemanticOperator.Int8Op.Ge | SemanticOperator.Int16Op.Ge | SemanticOperator.Int32Op.Ge
+               | SemanticOperator.Int64Op.Ge | SemanticOperator.BigIntOp.Ge => BinaryOperator.GreaterEqual
+
           case _ => ??? // TODO
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:447]: Is this case needed anymore?

If so, please change it to `throw InternalCompilerException("Unexpected: ...`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 2]
```diff
@@ -597,10 +622,10 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
           case "/" => mkApplyFqn("Div.div", List(e1, e2), sp1, sp2)
           case "%" => mkApplyFqn("Rem.rem", List(e1, e2), sp1, sp2)
           case "**" => mkApplyFqn("Exp.exp", List(e1, e2), sp1, sp2)
-          case "<" => WeededAst.Expression.BinaryDeprecated(BinaryOperator.Less, e1, e2, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:600]: I think UnaryDeprecated and BinaryDeprecated can now be removed from all ASTs.

**Review by magnus-madsen — COMMENTED** [review]
I guess once this is merged in, we can express the spaceship operator as a combination of lessEqual and greaterEqual?

**Comment by esbenbjerre** [comment]
> Does this mean that we can remove BinaryDeprecated? :)

I think unary plus is missing? :)

**Comment by esbenbjerre** [comment]
> I guess once this is merged in, we can express the spaceship operator as a combination of lessEqual and greaterEqual?

Yes. Another option is to add compare to Ord (Haskell does this) and make spaceship an alias for compare (you mentioned this earlier). What do you think?

**Comment by magnus-madsen** [comment]
> 
> 
> > Does this mean that we can remove BinaryDeprecated? :)
> 
> I think unary plus is missing? :)

I don't think Unary plus does anything. Does it do anything? Other than make the code line up when you have +1 and -1?

**Comment by magnus-madsen** [comment]
> 
> 
> > I guess once this is merged in, we can express the spaceship operator as a combination of lessEqual and greaterEqual?
> 
> Yes. ~Another option is to add `compare(a, a): Ordering` to `Ord` (Haskell does this) and make spaceship an alias for `compare` (you mentioned this earlier). What do you think?~
> EDIT: This would force users to implement identical `compare` functions for all types with `Ord` instances. Haskell solves this with compiler magic, such that a valid `Ord` instance is either `compare` or the relational operators.

Ultimately I think we want the following:

`Ord` has `compare` which returns an `Ordering` value. Then `<`, `<=` and so forth are defined in terms of `compare`.

For now, I want to merge in the current PR. 

In the next PR I think we should `compare` which returns an Ordering to `Ord`. Then we should rewrite the spaceship to call `compare`. Then we will have to update all TC instances to *also* implement `compare`. Thus there will be lots of duplicate, but that is fine for now, later we will then implement e.g. `<` in terms of compare and remove the duplicated code from the TCs.

**Pull request merged**